---
title: "Differential Expression Analysis for bulk RNA-seq data"
subtitle: "Vehicle Condition: CTRL vs FXS"
author: "Ximing Ran"
date: "2025-03-26"
output:
  pdf_document:
    toc: true
    toc_depth: 3
    number_sections: false
  html_document:
    # code_folding: hide
    toc: true
    toc_float: true
hitheme: tomorrow
highlighter: highlight.js

---

```{r setup, include=FALSE}
# load libraries
library(tidyverse)
library(knitr)
set.seed(2024)

knitr::opts_chunk$set(
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  fig.path = "./results/DEG-Veh_CTRL_vs_FXS/Analysis_figure/"
)

theme_set(theme_bw() + theme(legend.position = "bottom"))
knitr::kable(head(mtcars[, 1:4]), "simple")
```

```{r}
library(tibble)
library(tidyr)
library(dplyr)
library(rtracklayer)
```



```{r local_function_load}
# load function from local files
source(here::here("source", "DEG_functions.R"))
```

\newpage

# 1. Read the count data
In this section, we will read the clean count data from the synaptosomes_bulkRNA folder.  We will read the data and merge them into a single table. The final table will be stored in `../dataresults/bulkRNA_counts_clean.csv`.

```{r load_bulkRNA_data}
input_count <- read.csv(here::here("data", "bulkRNA",
                                      "bulkRNA_counts_cleaned.csv"))
counts <-  as.data.frame(input_count) %>% 
  column_to_rownames(var = "gene")
colnames(counts) <- gsub("_", "-", colnames(counts))

# if the colname is start with X, remove it
colnames(counts) <- gsub("^X", "", colnames(counts))


# raw sample list
sample_list_raw <- read.csv(here::here("data", "bulkRNA",
                                      "sample_info_FXS.csv")) %>%
                    mutate(condition = Label,
                           sample = gsub("_", "-", Sample_ID))


# Ensure the column names of counts exist in Sample.name
new_colnames <- sample_list_raw$sample[match(colnames(counts), sample_list_raw$sample )]

# Assign new column names
colnames(counts) <- new_colnames


# sort the columns by the colname
condition_list <- data.frame(
  group =sample_list_raw$condition
)

row.names(condition_list) <- sample_list_raw$sample

counts<- counts[, rownames(condition_list)]

gene_name_mapping<- readRDS(here::here("data","ref" ,"gene_name_mapping.rds"))




```


# 2. Differential expression analysis

In this section, we will perform differential expression analysis using DESeq2. 

```{r DESeq2_analysis}
# Init the result folder structure for the result
result_folder_all = './results/DEG-Veh_CTRL_vs_FXS'
result_folder = result_folder_all
Result_folder_structure(result_folder)

# load the comparison group information
reference_group <- "CTRL_VEH"
compare_group <- "FXS_VEH"
reference_group_short <- reference_group 
compare_group_short <- compare_group 


filter_sample_info <- condition_list %>% 
  filter(group %in% c(reference_group, compare_group))
filter_counts <- counts[, rownames(filter_sample_info)]

# Run the DESeq2 analysis
dds_obj <- DEAnalysis(counts =filter_counts, 
                       reference_group = reference_group,
                       compare_group = compare_group,
                       condition_list = filter_sample_info,
                       target_gene = target_gene,
                       result_folder = result_folder)


res <- results(dds_obj)
resOrdered <- res[order(res$padj), ]

# omit the NA values
resOrdered <- resOrdered[!is.na(resOrdered$padj),]
dds_obj <- dds_obj[rownames(resOrdered),]
write.csv(resOrdered, file.path(result_folder,"02-DEG", "01_all_gene_results.csv"))

# DEG with log2fc > 1 and padj < 0.05
deg_1 <- resOrdered %>% as.data.frame() %>% rownames_to_column(var = "gene") %>% 
  filter(padj < 0.05 & abs(log2FoldChange) > 1) %>% arrange(padj)
deg_1 <- deg_1[!is.na(deg_1$padj),]
write.csv(deg_1, file.path(result_folder,"02-DEG","02_DEG_log2fc_1.csv"), row.names = FALSE)

# DEG with log2fc > 1.5 and padj < 0.05
deg_1.5 <- resOrdered %>% as.data.frame() %>% rownames_to_column(var = "gene") %>% 
  filter(padj < 0.05 & abs(log2FoldChange) > 1.5) %>% arrange(padj)
deg_1.5  <- deg_1.5 [!is.na(deg_1.5 $padj),]
write.csv(deg_1.5 , file.path(result_folder,"02-DEG","03_DEG_log2fc_1_5.csv"), row.names = FALSE)
print("DEG analysis is done")
  
# Save the normalized counts
normalized_counts <- counts(dds_obj, normalized = TRUE)
write.csv(normalized_counts, file.path(result_folder,"02-DEG", "DESeq2_normalized_counts.csv"))

```


\newpage

# 3. Visualization for reuslt

## (1) Sample information
```{r Sample_dis_Heatmap, fig.width=8, fig.height=6}
# Plot the sample distance 
plot_sample_heatmap(dds_obj, figure_folder =
                    file.path(result_folder,"01-Sample_info"),
                    file_name = "01_sample_distance_heatmap")
```

\newpage
```{r  Sample_PCA, fig.width=8, fig.height=6}
# Plot the PCA plot for the sample
plot_sample_PCA_plot(dds_obj,figure_folder = file.path(result_folder,"01-Sample_info"),
                   file_name = "02_sample_PCA_plot",
                   reference_group, compare_group)
```

\newpage

## (2) DEG visualization - Volcano plot and Heatmap

```{r DEG_volcano_heatmap}
result_df <- results(dds_obj) %>%
  as.data.frame() %>%
  rownames_to_column(var = "GeneName") %>%
  dplyr::select(GeneName, everything()) %>%
  filter(!is.na(padj)) %>%  # Correct way to filter non-NA values
  arrange(padj)
```

```{r volcano_DEG_fc_1, fig.width=6, fig.height=6}
plot_volcano_plot(result_df=result_df, 
                  figure_folder = file.path(result_folder,"02-DEG"),
                  file_name = "02_volcano_plot_log2fc_1",
                  
                  # max_p = 3,max_fc = 5,
                  thread = 1 , dot_size =2, label_gene = NULL)


```

\newpage
```{r volcano_DEG_fc_1.5, fig.width=6, fig.height=6}
plot_volcano_plot(result_df=result_df, 
                  figure_folder = file.path(result_folder,"02-DEG"),
                  file_name = "02_volcano_plot_log2fc_1.5",
                  # max_p = 3,max_fc = 5,
                  thread = 1.5 ,  dot_size =2,label_gene = NULL)

```

\newpage

```{r volcano_DEG_2, fig.width=6, fig.height=6}
# Plot the volcano plot for the DEG
plot_volcano_plot(result_df=result_df, 
                  figure_folder = file.path(result_folder,"02-DEG"),
                  file_name = "02_volcano_plot_log2fc_2",
                  # max_p = 3,max_fc = 5,
                  thread = 2 ,  dot_size =2,label_gene = NULL)
```

\newpage

```{r heatmap_DEG, fig.width=8, fig.height=8}
# Plot the heatmap for the DEG
vsd_obj <- varianceStabilizingTransformation(dds_obj, blind = TRUE)

DEG_gene_2 <- result_df %>% filter(abs(log2FoldChange) > 2 & padj <0.05) %>% pull(GeneName)
DEG_gene_1.5 <- result_df %>% filter(abs(log2FoldChange) > 1.5 & padj <0.05) %>% pull(GeneName)
DEG_gene_1 <- result_df %>% filter(abs(log2FoldChange) > 1 & padj <0.05) %>%  pull(GeneName)
```



```{r heatmap_DEG_fc_1, fig.width=8, fig.height=8}
plot_gene_heatmap(vsd_obj, gene_list = DEG_gene_1, 
                  figure_folder = file.path(result_folder,"02-DEG"),
                  file_name = "03_heatmap_log2fc_1",
                  reference_group, compare_group,
                  cluster_rows = TRUE, cluster_cols = FALSE, 
                  scale = "none")
```

\newpage
```{r heatmap_DEG_fc_1_row, fig.width=8, fig.height=8}
plot_gene_heatmap(vsd_obj, gene_list = DEG_gene_1, 
                  figure_folder = file.path(result_folder,"02-DEG"),
                  file_name = "03_heatmap_log2fc_1_row",
                  reference_group, compare_group,
                  cluster_rows = TRUE, cluster_cols = FALSE, 
                  scale = "row")
```



\newpage

```{r heatmap_DEG_fc_1.5, fig.width=8, fig.height=8}
plot_gene_heatmap(vsd_obj, gene_list = DEG_gene_1.5, 
                  figure_folder = file.path(result_folder,"02-DEG"),
                  file_name = "03_heatmap_log2fc_1.5",
                  reference_group, compare_group,
                  cluster_rows = TRUE, cluster_cols = FALSE, 
                  scale = "none")
```

\newpage
```{r heatmap_DEG_fc_1.5_row, fig.width=8, fig.height=8}
plot_gene_heatmap(vsd_obj, gene_list = DEG_gene_1.5, 
                  figure_folder = file.path(result_folder,"02-DEG"),
                  file_name = "03_heatmap_log2fc_1.5_row",
                  reference_group, compare_group,
                  cluster_rows = TRUE, cluster_cols = FALSE, 
                  scale = "row")
```



```{r heatmap_DEG_fc_2, fig.width=8, fig.height=8}
plot_gene_heatmap(vsd_obj, gene_list = DEG_gene_2, 
                  figure_folder = file.path(result_folder,"02-DEG"),
                  file_name = "03_heatmap_log2fc_2",
                  reference_group, compare_group,
                  cluster_rows = TRUE, cluster_cols = FALSE, 
                  scale = "none")
```


\newpage

```{r heatmap_DEG_fc_2_row, fig.width=8, fig.height=8}
plot_gene_heatmap(vsd_obj, gene_list = DEG_gene_2, 
                  figure_folder = file.path(result_folder,"02-DEG"),
                  file_name = "03_heatmap_log2fc_2_row",
                  reference_group, compare_group,
                  cluster_rows = TRUE, cluster_cols = FALSE, 
                  scale = "row")
```


\newpage

# 4. GSVA analysis

```{r GSVA_analysis, fig.width=4, fig.height=6}

# # The following code is used to generate the GSVA matrix , only need to run once
# gmxFile <- here::here("data", "ref", "c5.go.v2023.1.Hs.symbols.gmt")
# go_list <- getGmt(gmxFile)
# 
# geneset <- go_list
# dat <- as.matrix(counts)
# 
# gsvapar <- gsvaParam(dat, geneset, maxDiff=TRUE)
# gsva_es <- gsva(gsvapar)
# gsva_matrix <- as.data.frame(gsva_es)
# 
# # save the result
# # write.csv(gsva_matrix, file.path(result_folder,"04-GSVA", "01_GSVA_matrix.csv"))
# write.csv(gsva_matrix, "./results/Mutil_Group_GSVA/01_GSVA_matrix.csv")

gsva_matrix <- read.csv("./results/Mutil_Group_GSVA/01_GSVA_matrix.csv",
                        row.names = 1)

colnames(gsva_matrix) <- sub("^X", "", colnames(gsva_matrix))
colnames(gsva_matrix) <- gsub("\\.", "-", colnames(gsva_matrix))

condition_list_label <- condition_list %>%
  filter(group %in% c(reference_group, compare_group)) %>%
  mutate(group = case_when(
    group == reference_group ~ reference_group_short,
    group == compare_group ~ compare_group_short,
    TRUE ~ group  # fallback just in case
  ))


# plot the heatmap for the GSVA result
pathway_list <- read.csv(here::here("data", "ref", "focus-pathway_2024_10_03.csv"))


# # plot for all pathway
# for (i in 1:nrow(pathway_list)){
#   if (i %% 10 == 0) print(i)
#   pathway_name <- pathway_list$pathway[i]
#   plot_gsva_boxplot(gsva_matrix,
#                     condition_list_label =condition_list_label,
#                     pathway_name =  pathway_name,
#                     figure_folder = file.path(result_folder,"04-GSVA","Boxplot"),
#                     file_name = paste0("GSVA_", pathway_name),
#                     fig.height = 6, fig.width = 4,
#                     reference_group =reference_group_short ,
#                     compare_group = compare_group_short)
# 
# }

box_plot_folder<- file.path(result_folder,"04-GSVA","Boxplot")
# create the folder
dir.create(box_plot_folder, showWarnings = FALSE)


# plot for the focus pathway
for (i in 1:2){
  pathway_name <- pathway_list$pathway[i]
  print(pathway_name)
  p<-plot_gsva_boxplot(gsva_matrix,
                    condition_list_label =condition_list_label,
                    pathway_name =  pathway_name,
                    figure_folder = file.path(result_folder,"04-GSVA","Boxplot"),
                    file_name = paste0("GSVA_", pathway_name),
                    fig.height = 6, fig.width = 4,
                    reference_group =reference_group_short ,
                    compare_group = compare_group_short)
  print(p)
}


```

\newpage

# 5. Pathway Enrichment Analysis

## (1) DEG threshold: log2fc > 1

```{r Enrichment_DEG_1,fig.width=12, fig.height=8}
deg1 <- result_df %>% filter(padj < 0.05 & abs(log2FoldChange) > 1)
up_gene_1 <- deg1 %>% filter(log2FoldChange > 0) %>% pull(GeneName)
down_gene_1 <- deg1 %>% filter(log2FoldChange < 0) %>% pull(GeneName)


# test for the function
if (length(up_gene_1) == 0) {
  print("The up_gene_1 is empty, skip the analysis")
} else {
  Enrichment_analysis(gene_list = up_gene_1,
                      result_folder = file.path(result_folder,"03-Enrichment"),
                      file_name = "02-DEG_1_up", gene_name_mapping, flag = "Up")
}

if (length(down_gene_1) == 0) {
  print("The down_gene_1 is empty, skip the analysis")
} else {
  Enrichment_analysis(gene_list = down_gene_1,
                      result_folder = file.path(result_folder,"03-Enrichment"),
                      file_name = "02-DEG_1_down", gene_name_mapping, flag = "Down")
}



```

\newpage

## (2) DEG threshold: log2fc > 1.5

```{r Enrichment_DEG_1.5,fig.width=12, fig.height=8}
deg1.5 <- result_df %>% filter(padj < 0.05 & abs(log2FoldChange) > 1.5)
up_gene_1.5 <- deg1.5 %>% filter(log2FoldChange > 0) %>% pull(GeneName)
down_gene_1.5 <- deg1.5 %>% filter(log2FoldChange < 0) %>% pull(GeneName)

# test for the function
# if the up_gene_1.5 is empty, we will skip the analysis
if (length(up_gene_1.5) == 0) {
  print("The up_gene_0.5 is empty, skip the analysis")
} else {
  Enrichment_analysis(gene_list = up_gene_1.5,
                      result_folder = file.path(result_folder,"03-Enrichment"),
                      file_name = "02-DEG_1.5_up", gene_name_mapping, flag = "Up")
}

if (length(down_gene_1.5) == 0) {
  print("The down_gene_1.5 is empty, skip the analysis")
} else {
  Enrichment_analysis(gene_list = down_gene_1.5,
                      result_folder = file.path(result_folder,"03-Enrichment"),
                      file_name = "02-DEG_1.5_down", gene_name_mapping, flag = "Down")
}
```


\newpage

## (3) DEG threshold: log2fc > 1.5

```{r Enrichment_DEG_2,fig.width=12, fig.height=8}
deg2<- result_df %>% filter(padj < 0.05 & abs(log2FoldChange) > 2)
up_gene_2 <- deg2 %>% filter(log2FoldChange > 0) %>% pull(GeneName)
down_gene_2 <- deg2 %>% filter(log2FoldChange < 0) %>% pull(GeneName)

# test for the function
# if the up_gene_1.5 is empty, we will skip the analysis
if (length(up_gene_2) == 0) {
  print("The up_gene_2 is empty, skip the analysis")
} else {
  Enrichment_analysis(gene_list = up_gene_2,
                      result_folder = file.path(result_folder,"03-Enrichment"),
                      file_name = "02-DEG_2_up", gene_name_mapping, flag = "Up")
}

if (length(down_gene_2) == 0) {
  print("The down_gene_2 is empty, skip the analysis")
} else {
  Enrichment_analysis(gene_list = down_gene_2,
                      result_folder = file.path(result_folder,"03-Enrichment"),
                      file_name = "02-DEG_2_down", gene_name_mapping, flag = "Down")
}
```


\newpage
# Session information
```{r}
sessionInfo()
```

